api/rest/coordinator.go

```diff
@@ -0,0 +1,256 @@
+// across a cluster of nodes
+type Coordinator interface {
+	// CreateBundle starts the bundle creation process. Status updates be monitored
+	// on the returned channel.
+	CreateBundle(ctx context.Context, id string, nodes []node) <-chan BundleStatus
+	// CollectBundle waits until all the nodes' bundles have finished, downloads,
+	// and merges them. The resulting bundle zip file path is returned.
+	CollectBundle(ctx context.Context, bundleID string, numBundles int, statuses <-chan BundleStatus) (string, error)
+}
+
+// ParallelCoordinator implements Coordinator interface to coordinate bundle
+// creation across a cluster, parallelized across multiple goroutines
+type ParallelCoordinator struct {
+	client              Client
+	statusCheckInterval time.Duration
+	workDir             string
+}
+
+// job is a function that will be called by the worker function. The output will be added to results channel
+type job func() BundleStatus
+
+// worker is a function that will run incoming jobs from jobs channel and put jobs output to results chan
+func worker(ctx context.Context, jobs <-chan job, results chan<- BundleStatus) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case j := <-jobs:
+			results <- j()
+		}
+	}
+}
+
+// NewParallelCoordinator constructs a new parallelcoordinator.
+func NewParallelCoordinator(client Client, interval time.Duration, workDir string) ParallelCoordinator {
+	return ParallelCoordinator{
+		client:              client,
+		statusCheckInterval: interval,
+		workDir:             workDir,
+	}
+}
+
+// CreateBundle starts the bundle creation process. Status updates be monitored
+// on the returned channel.
+func (c ParallelCoordinator) CreateBundle(ctx context.Context, id string, nodes []node) <-chan BundleStatus {
+
+	jobs := make(chan job)
+	statuses := make(chan BundleStatus)
+
+	for i := 0; i < numberOfWorkers; i++ {
+		go worker(ctx, jobs, statuses)
+	}
+
+	for _, n := range nodes {
+		logrus.WithField("IP", n.IP).Info("Sending creation request to node.")
+
+		// necessary to prevent the closure from giving the same node to all the calls
+		tmpNode := n
+		jobs <- func() BundleStatus {
+			select {
+			case <-ctx.Done():
+				return BundleStatus{
+					id:   id,
+					node: tmpNode,
+					err:  fmt.Errorf(contextDoneErrMsg),
+				}
+			default:
+			}
+
+			return c.createBundle(ctx, tmpNode, id, jobs)
+		}
+	}
+
+	return statuses
+}
+
+// CollectBundle waits until all the nodes' bundles have finished, downloads,
+// and merges them. The resulting bundle zip file path is returned.
+func (c ParallelCoordinator) CollectBundle(ctx context.Context, bundleID string, numBundles int, statuses <-chan BundleStatus) (string, error) {
+
+	// holds the paths to the downloaded local bundles before merging
+	var bundlePaths []string
+
+	finishedBundles := 0
+	for finishedBundles < numBundles {
+		select {
+		case <-ctx.Done():
+			return "", fmt.Errorf("context canceled before all node bundles finished")
```

Comment:
I'm not sure if this is a valid approach. What if we have problems with a single node that is not responding? Then a whole bundle will not be generated. I think we should return generated bundle with information that timeout occured

Reply:
I think this is best handled as part of https://jira.mesosphere.com/browse/DCOS_OSS-5303. In the meantime, I'm going to have it merge the zips that have already been downloaded and log that the context was cancelled.